### PR TITLE
TorchVision

### PR DESCRIPTION
**Pull request by Rob1in**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+| `input size`                 | List[int]             | Optional     | `None`      | Resize the image. Overides resize from weights metadata.                                                                                                                                                                                                                                                                                           |
+| `mean_rgb`                   | [float, float, float] | Optional     | `[0, 0, 0]` | Specifies the mean and standard deviation values for normalization in RGB order                                                                                                                                                                                                                                                                    |
```
Comment (bhaney) [inline README.md]: For all of the optional configs, i thought most torchvision models already applied the correct normalization and such? Or is that not the case, and the user must know the preprocessing for the image and input them into the config? Or are these just for over-writing the recommendation already in torchvision?

File: src/test_local.py [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+from viam.utils import dict_to_struct
+import asyncio
+import numpy as np
+
+
+async def main():
+    tvs = TorchVisionService(name="torch_vision_service")
+    # tvs.model = get_model("resnet18")
+    cfg = ServiceConfig(attributes=dict_to_struct({"model_name": "resnet50"}))
+    # cfg.attributes.fields["model_name"] = "resnet18"
```
Comment (bhaney) [inline src/test_local.py]: remove the debug comments

File: src/test_local.py [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+
+    tvs.reconfigure(config=cfg, dependencies=None)
+    path_to_input_image = "/Users/robinin/torch-infer/torchvision/src/grass_hopper.jpg"
+    input_image = np.array(Image.open(path_to_input_image))
+    # input_image = torchvision.io.read_image(path_to_input_image)
```
Comment (bhaney) [inline src/test_local.py]: remove this comment

File: src/test_local.py [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+
+
+async def main():
+    tvs = TorchVisionService(name="torch_vision_service")
+    # tvs.model = get_model("resnet18")
+    cfg = ServiceConfig(attributes=dict_to_struct({"model_name": "resnet50"}))
+    # cfg.attributes.fields["model_name"] = "resnet18"
+
+    tvs.reconfigure(config=cfg, dependencies=None)
+    path_to_input_image = "/Users/robinin/torch-infer/torchvision/src/grass_hopper.jpg"
```
Comment (bhaney) [inline src/test_local.py]: Don't refer to your personal paths but replace with a place holder

File: src/torchvision_module.py [hunk 5]
```diff
@@ -0,0 +1,291 @@
...
+        extra: Optional[Mapping[str, Any]] = None,
+        timeout: Optional[float] = None,
+    ) -> List[Classification]:
+        if not self.properties.implements_classification:
+            return NotImplementedError
+
+        input_tensor = self.preprocessor(image)
+        # input_tensor = image
+        # torchvision.utils.save_image(input_tensor, "./input.jpg")
+        # img = decode_image(image)
```
Comment (bhaney) [inline src/torchvision_module.py]: remove comments

File: src/torchvision_module.py [hunk 6]
```diff
@@ -0,0 +1,291 @@
...
+        with torch.no_grad():
+            prediction: Tensor = self.model(input_tensor)
+
+        prediction = prediction.squeeze(0)
+        prediction = prediction.softmax(0)
```
Comment (bhaney) [inline src/torchvision_module.py]: It's nice that torchvision all follow the same conventions. You wouldn't have been able to assume this in the mlmodel service

File: src/torchvision_module.py [hunk 7]
```diff
@@ -0,0 +1,291 @@
...
+            self.weights.meta["categories"][index] for index in top_indices
+        ]
+        res = [
+            Classification(class_name=name, confidence=score)
+            for name, score in zip(category_names, scores)
+        ]
+        res = self.filter_output(res)
+        return res
+
+    async def get_classifications_from_camera(
```
Comment (bhaney) [inline src/torchvision_module.py:169]: There is a lot of repeated code here. Could you put the repeated parts of the function in another function, and have both get_classifications and get_classifications_from_camera call it

File: src/torchvision_module.py [hunk 8]
```diff
@@ -0,0 +1,291 @@
...
+            self.weights.meta["categories"][index] for index in top_indices
+        ]
+        res = [
+            Classification(class_name=name, confidence=score)
+            for name, score in zip(category_names, scores)
+        ]
+        res = self.filter_output(res)
+        return res
+
+    async def get_detections_from_camera(
```
Comment (bhaney) [inline src/torchvision_module.py:186]: same with here - don't repeat code, but put the shared code in a third function. In that way, if you ever have to edit this code, you can avoid a bug where you change it in one function, but not another

File: src/utils.py [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+
+
+def decode_image(image: Union[Image.Image, RawImage, ViamImage]) -> np.ndarray:
+    """decode image to BGR numpy array
+
+    Args:
+        raw_image (Union[Image.Image, RawImage])
+
+    Returns:
+        np.ndarray: BGR numpy array
```
Comment (bhaney) [inline src/utils.py]: Is this comment correct? Are you converting to BGR, or RGB?

File: src/torchvision_module.py [hunk 10]
```diff
@@ -0,0 +1,291 @@
...
+        self, config: ServiceConfig, dependencies: Mapping[ResourceName, ResourceBase]
+    ):
+        self.dependencies = dependencies
+        device = (
+            torch.device("cuda") if torch.cuda.is_available() else torch.device("cpu")
+        )
+
+        # self.camera_name = config.attributes.fields["camera_name"].string_value
+
+        # self.camera = dependencies[Camera.get_resource_name(self.camera_name)]
```
Comment (bhaney) [inline src/torchvision_module.py]: can remove these comments

**Review by bhaney — CHANGES_REQUESTED** [review]
Mostly just clean up, and some refactoring

File: README.md (hunk 1 above)
Comment (Rob1in) [reply README.md]: Torchvision don't guarantee that metadata would always carry preprocessing information. Also because they expect the community to add models. I made it more explicit in the README and pushed those optional configs at the very end.

File: src/test_local.py (hunk 2 above)
Comment (Rob1in) [reply src/test_local.py]: done

File: src/test_local.py (hunk 3 above)
Comment (Rob1in) [reply src/test_local.py]: done

File: src/test_local.py (hunk 4 above)
Comment (Rob1in) [reply src/test_local.py]: done

File: src/torchvision_module.py (hunk 5 above)
Comment (Rob1in) [reply src/torchvision_module.py]: done

File: src/torchvision_module.py (hunk 7 above)
Comment (Rob1in) [reply src/torchvision_module.py:169]: done

File: src/torchvision_module.py (hunk 8 above)
Comment (Rob1in) [reply src/torchvision_module.py:186]: done

File: src/torchvision_module.py (hunk 10 above)
Comment (Rob1in) [reply src/torchvision_module.py]: done

**Review by bhaney — APPROVED** [review]
LGTM!
